jupyterhub_traefik_proxy/redis.py

```diff
@@ -0,0 +1,166 @@
+        url = urlparse(self.redis_url)
+        redis_config = {
+            "endpoints": [url.netloc],
+            "rootKey": self.kv_traefik_prefix,
+        }
+        if self.redis_username:
+            redis_config["username"] = self.redis_username
+        if self.redis_password:
+            redis_config["password"] = self.redis_password
+
+        self.static_config = deep_merge(
+            self.static_config, {"providers": {"redis": redis_config}}
+        )
+        return super()._setup_traefik_static_config()
+
+    async def _kv_atomic_set(self, to_set: dict):
+        """Set a collection of keys and values
+
+        Should be done atomically (i.e. in a transaction),
+        setting nothing on failure.
+
+        Args:
+
+        to_set (dict): key/value pairs to set
+            Will always be a flattened dict
+            of single key-value pairs,
+            not a nested structure.
+        """
+        self.log.debug("Setting redis keys %s", to_set.keys())
+        await self.redis.mset(to_set)
+
+    _delete_script = Any()
+
+    @default("_delete_script")
+    def _register_delete_script(self):
+        """Register LUA script for deleting all keys matching in a prefix
+
+        Doing the scan & delete from Python is _extremely_ slow
+        for some reason
+        """
+        _delete_lua = """
+        local all_keys = {};
+        local cursor = "";
+        repeat
+            local result = redis.call("SCAN", cursor, "match", ARGV[1], "count", ARGV[2])
+            cursor = result[1];
+            for i, key in ipairs(result[2]) do
+                table.insert(all_keys, key);
+            end
+        until cursor == "0"
+        for i, key in ipairs(all_keys) do
+            redis.call("DEL", key);
+        end
+        return #all_keys;
+        """
+        return self.redis.register_script(_delete_lua)
+
+    async def _kv_atomic_delete(self, *keys):
+        """Delete one or more keys
+
+        If a key ends with `self.kv_separator`, it should be a recursive delete
+        """
+
+        to_delete = []
+
+        futures = []
+        for key in keys:
+            if key.endswith(self.kv_separator):
+                prefix = key + "*"
+                self.log.debug("Deleting redis tree %s", prefix)
+                f = asyncio.ensure_future(self._delete_script(args=[prefix, 100]))
+                f.add_done_callback(
+                    lambda f: self.log.debug(
+                        "Deleted %i keys in %s", f.result(), prefix
+                    )
+                )
+                futures.append(f)
+            else:
+                to_delete.append(key)
+
+        if to_delete:
+            self.log.debug("Deleting redis keys %s", to_delete)
+            futures.append(self.redis.delete(*to_delete))
+
+        await asyncio.gather(*futures)
+
+    async def _kv_get_tree(self, prefix):
+        """Return all data under prefix as a dict
+
+        Should probably use `unflatten_dict_from_kv`
```

Comment:
I see that `unflatten_dict_from_kv` is already used here, so maybe it's ok for this comment to go?

Reply:
I'll add a commit about this in #231!